src/common/serializer/buffered_file_writer.cpp

```diff
@@ -22,16 +23,34 @@ idx_t BufferedFileWriter::GetTotalWritten() {
 }
 
 void BufferedFileWriter::WriteData(const_data_ptr_t buffer, idx_t write_size) {
-	// first copy anything we can from the buffer
-	const_data_ptr_t end_ptr = buffer + write_size;
-	while (buffer < end_ptr) {
-		idx_t to_write = MinValue<idx_t>((end_ptr - buffer), FILE_BUFFER_SIZE - offset);
-		D_ASSERT(to_write > 0);
-		memcpy(data.get() + offset, buffer, to_write);
-		offset += to_write;
-		buffer += to_write;
-		if (offset == FILE_BUFFER_SIZE) {
-			Flush();
+	if (write_size >= (2 * FILE_BUFFER_SIZE - offset)) {
+		idx_t to_copy = 0;
+		// Check before performing direct IO if there is some data in the current internal buffer.
+		// If so, then fill the buffer (to avoid to small write operation), flush it and then write
+		// all the remain data directly.
+		// This is to avoid to split a large buffer into N*FILE_BUFFER_SIZE buffers
+		if (offset != 0) {
+			// Some data are still present in the buffer let write them before
+			to_copy = FILE_BUFFER_SIZE - offset;
+			memcpy(data.get() + offset, buffer, to_copy);
+			offset += to_copy;
+			Flush(); // Flush buffer before writing every things else
```

Comment:
Can we call `Flush()` followed by `fs.Write()` of the entire incoming buffer below without first copying part of the buffer into the existing buffer? I don't quite see the benefit of first copying part of the buffer into the original file buffer, and it complicates the code here. Two write calls are happening in both cases.

Reply:
In my opinion the underlying FS expect that the buffered are size at least to the size that has been configured. On a remote FS I think it may impact some performance (depending of course of the implementation), and I wanted to avoid the underlying FS to add an extract layer of buffers.
This is only my opinion and you have the final words regarding this. Please confirm if you want me to do the change or not :)

Reply:
I can see that - let's keep it as-is then